File: CS/Tests/CRUDTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using Default;
using NUnit.Framework;
using WebApplication1.Models;

namespace Tests
{
    [TestFixture]
    public class CRUDOperationTests : ODataTestsBase {

        [Test]
        public void CreateObject() {
            Container container = GetODataContainer();
            Customer customer = new Customer() {
                CustomerID = "C0001",
                CompanyName = "Test Company"
            };
            container.AddToCustomer(customer);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.Created, response.First().StatusCode);

            container = GetODataContainer();
            Customer createdItem = container.Customer.Where(t => t.CustomerID == customer.CustomerID).First();
            Assert.AreEqual(customer.CustomerID, createdItem.CustomerID);
            Assert.AreEqual(customer.CompanyName, createdItem.CompanyName);
        }

        [Test]
        public void UpdateObject() {
            Container container = GetODataContainer();
            Customer customer = container.Customer.Where(t => t.CustomerID == "BSBEV").First();

            customer.CompanyName = "Test Company Renamed";
            container.UpdateObject(customer);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Customer updatedItem = container.Customer.Where(t => t.CustomerID == "BSBEV").First();
            Assert.AreEqual(customer.CustomerID, updatedItem.CustomerID);
            Assert.AreEqual(customer.CompanyName, updatedItem.CompanyName);
        }

        [Test]
        public void BatchUpdate() {
            Container container = GetODataContainer();
            var customers = container.Customer.ToList();
            foreach(var customer in customers) {
                customer.CompanyName += " -renamed";
                container.UpdateObject(customer);
            }
            var response = container.SaveChanges(Microsoft.OData.Client.SaveChangesOptions.BatchWithSingleChangeset);

            Assert.AreEqual(4, response.Count());
            foreach(var res in response) {
                Assert.AreEqual((int)HttpStatusCode.NoContent, res.StatusCode);
            }

            container = GetODataContainer();
            customers = container.Customer.ToList();
            foreach(var customer in customers) {
                Assert.IsTrue(customer.CompanyName.EndsWith(" -renamed"));
            }
        }

        [Test]
        public void DeleteObject() {
            Container container = GetODataContainer();
            Customer customer = container.Customer.Where(t => t.CustomerID == "WARTH").First();

            container.DeleteObject(customer);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);
        }

        [Test]
        public void CreateObjectWithBlobField() {
            byte[] pictureData = new byte[256 * 256 * 4];
            for(int i = 0; i < pictureData.Length; i++) {
                pictureData[i] = (byte)i;
            }
            Container container = GetODataContainer();
            Product product = new Product() {
                ProductName = "test product",
                Picture = pictureData
            };
            container.AddToProduct(product);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.Created, response.First().StatusCode);

            container = GetODataContainer();
            Product createdItem = container.Product.Where(t => t.ProductName == product.ProductName).First();

            Assert.IsNotNull(createdItem.Picture);
            Assert.AreEqual(product.Picture.Length, createdItem.Picture.Length);
        }

        [Test]
        public void UpdateObjectWithBlobField() {
            byte[] pictureData = new byte[256 * 256 * 4];
            for(int i = 0; i < pictureData.Length; i++) {
                pictureData[i] = (byte)i;
            }
            Container container = GetODataContainer();
            var product = container.Product.Where(t => t.ProductName == "Queso Cabrales").First();
            byte[] oldPicture = product.Picture;
            product.Picture = pictureData;
            container.UpdateObject(product);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Product updatedItem = container.Product.Where(t => t.ProductID == product.ProductID).First();
            Assert.IsNull(oldPicture);
            Assert.IsNotNull(updatedItem.Picture);
            Assert.AreEqual(pictureData.Length, updatedItem.Picture.Length);
        }

        [Test]
        public void AddRelatedObject() {
            Container container = GetODataContainer();
            Order order = container.Order.OrderByDescending(t => t.ID).First();

            OrderDetail detail = new OrderDetail() {
                Quantity = 105,
                UnitPrice = 201.37m
            };
            container.AddRelatedObject(order, "OrderDetails", detail);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.Created, response.First().StatusCode);

            container = GetODataContainer();
            OrderDetail createdItem = container.OrderDetail.OrderByDescending(t => t.OrderDetailID).First();
            Assert.AreEqual(detail.Quantity, createdItem.Quantity);
            Assert.AreEqual(detail.UnitPrice, createdItem.UnitPrice);
            Assert.Greater(createdItem.OrderDetailID, 0);
        }

        [Test]
        public void UpdateRelatedObject() {
            Container container = GetODataContainer();
            Order order = container.Order.Expand(t => t.OrderDetails).OrderBy(t => t.Date).First();
            OrderDetail detail = order.OrderDetails.OrderBy(t => t.OrderDetailID).First();

            short oldQuantity = detail.Quantity;
            detail.Quantity += 1;
            container.UpdateObject(detail);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            OrderDetail updatedItem = container.Order.Expand(t => t.OrderDetails)
                .Where(t => t.ID == order.ID).First()
                .OrderDetails.Where(d => d.OrderDetailID == detail.OrderDetailID).First();
            Assert.AreNotEqual(oldQuantity, updatedItem.Quantity);
            Assert.AreEqual(detail.Quantity, updatedItem.Quantity);
        }

        [Test]
        public void DeleteRelatedObject() {
            Container container = GetODataContainer();
            Order order = container.Order.Expand(t => t.OrderDetails).OrderBy(t => t.Date).First();
            OrderDetail detail = order.OrderDetails.OrderBy(t => t.OrderDetailID).First();

            container.DeleteLink(order, "OrderDetails", detail);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Order updatedItem = container.Order.Expand(t => t.OrderDetails).Where(t => t.ID == order.ID).First();
            Assert.False(updatedItem.OrderDetails.Any(d => d.OrderDetailID == detail.OrderDetailID));
        }

        [Test]
        public void DeleteAggregatedCollection() {
            Container container = GetODataContainer();
            Order order = container.Order.Expand(t => t.OrderDetails).OrderBy(t => t.Date).First();

            container.DeleteObject(order);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            var details = container.OrderDetail.Where(t => t.Order.ID == order.ID).ToList();
            Assert.AreEqual(0, details.Count);
        }

        [Test]
        public void AddLink() {
            Container container = GetODataContainer();
            Order order = container.Order.Where(t => t.Customer == null).OrderByDescending(t => t.ID).First();
            Customer customer = container.Customer.Where(t => t.CustomerID == "BSBEV").First();

            container.AddLink(customer, "Orders", order);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Order updatedItem = container.Order.Expand(t => t.Customer).Where(t => t.ID == order.ID).First();
            Assert.IsNull(order.Customer);
            Assert.NotNull(updatedItem.Customer);
            Assert.AreEqual(updatedItem.Customer.CustomerID, customer.CustomerID);
        }

        [Test]
        public void UpdateLink() {
            Container container = GetODataContainer();
            Order order = container.Order.Where(t => t.Customer.CustomerID == "BSBEV").OrderByDescending(t => t.ID).First();
            Customer customer = container.Customer.Where(t => t.CustomerID == "SANTG").First();

            container.AddLink(customer, "Orders", order);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Order updatedItem = container.Order.Expand(t => t.Customer).Where(t => t.ID == order.ID).First();
            Assert.AreEqual(updatedItem.Customer.CustomerID, customer.CustomerID);
        }

        [Test]
        public void DeleteLink() {
            Container container = GetODataContainer();
            Order order = container.Order.Expand(t => t.Customer).Where(t => t.Customer.CustomerID == "BSBEV").OrderByDescending(t => t.ID).First();

            container.DeleteLink(order.Customer, "Orders", order);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Order updatedItem = container.Order.Expand(t => t.Customer).Where(t => t.ID == order.ID).First();
            Assert.IsNotNull(order.Customer);
            Assert.IsNull(updatedItem.Customer);
        }

        [Test]
        public void DeleteInheritedObject() {
            Container container = GetODataContainer();
            int contractId = container.Contract.Where(t => t.Number == "2018-0003").First().ID;
            BaseDocument contract = container.BaseDocument.Where(t => t.ID == contractId).First();

            container.DeleteObject(contract);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            int count = container.BaseDocument.ToList().Count(t => t.ID == contractId);
            Assert.AreEqual(0, count);
        }

        [Test]
        public void AddLinkToInherited() {
            Container container = GetODataContainer();
            Order order = container.Order.Where(t => t.ParentDocument == null).OrderByDescending(t => t.ID).First();
            Contract contract = container.Contract.Where(t => t.Number == "2018-0003").First();

            container.AddLink(contract, "LinkedDocuments", order);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Order updatedItem = container.Order.Expand(t => t.ParentDocument).Where(t => t.ID == order.ID).First();
            Assert.IsNull(order.ParentDocument);
            Assert.NotNull(updatedItem.ParentDocument);
            Assert.AreEqual(updatedItem.ParentDocument.GetType(), typeof(Contract));
            Assert.AreEqual(updatedItem.ParentDocument.ID, contract.ID);
        }

        [Test]
        public void UpdateLinkToInherited() {
            Container container = GetODataContainer();
            Order order = container.Order.Where(t => (t.ParentDocument as Contract).Number == "2018-0002").First();
            Contract contract = container.Contract.Where(t => t.Number == "2018-0003").First();

            container.AddLink(contract, "LinkedDocuments", order);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Order updatedItem = container.Order.Expand(t => t.ParentDocument).Where(t => t.ID == order.ID).First();
            Assert.AreEqual(updatedItem.ParentDocument.ID, contract.ID);
        }

        [Test]
        public void DeleteLinkToInherited() {
            Container container = GetODataContainer();
            Order order = container.Order.Expand(t => t.ParentDocument).Where(t => (t.ParentDocument as Contract).Number == "2018-0002").First();

            container.DeleteLink(order.ParentDocument, "LinkedDocuments", order);
            var response = container.SaveChanges();

            Assert.AreEqual(1, response.Count());
            Assert.AreEqual((int)HttpStatusCode.NoContent, response.First().StatusCode);

            container = GetODataContainer();
            Order updatedItem = container.Order.Expand(t => t.ParentDocument).Where(t => t.ID == order.ID).First();
            Assert.IsNotNull(order.ParentDocument);
            Assert.IsNull(updatedItem.ParentDocument);
        }

        [Test]
        public void SetLink() {
            Container container = GetODataContainer();
            Order order = new Order() {
                Date = DateTimeOffset.Now,
                OrderStatus = OrderStatus.New
            };
            container.AddToOrder(order);
            var response1 = container.SaveChanges();

            Assert.AreEqual(1, response1.Count());
            Assert.AreEqual((int)HttpStatusCode.Created, response1.First().StatusCode);
            Customer customer = new Customer() {
                CustomerID = "Duffy",
                CompanyName = "Acme"
            };
            container.AddToCustomer(customer);
            order.Customer = customer;
            container.SetLink(order, "Customer", customer);
            var response2 = container.SaveChanges();
            Assert.AreEqual(2, response2.Count());
            Assert.IsTrue(response2.Any(r => r.StatusCode == (int)HttpStatusCode.Created));
            Assert.IsTrue(response2.Any(r => r.StatusCode == (int)HttpStatusCode.NoContent));

            container = GetODataContainer();
            Order createdItem = container.Order.Expand(t => t.Customer).Where(t => t.ID == order.ID).First();
            Assert.IsNotNull(createdItem.Customer);
            Assert.AreEqual(order.Customer.CustomerID, createdItem.Customer.CustomerID);
        }


        [Test]
        public void RemoveLink() {
            Container container = GetODataContainer();
            Order order = new Order() {
                Date = DateTimeOffset.Now,
                OrderStatus = OrderStatus.New
            };
            container.AddToOrder(order);
            Customer customer = new Customer() {
                CustomerID = "Duffy",
                CompanyName = "Acme"
            };
            container.AddToCustomer(customer);
            container.SetLink(order, "Customer", customer);
            var response1 = container.SaveChanges();
            Assert.AreEqual(3, response1.Count());
            Assert.AreEqual(2, response1.Count(r => r.StatusCode == (int)HttpStatusCode.Created));
            Assert.IsTrue(response1.Any(r => r.StatusCode == (int)HttpStatusCode.NoContent));

            container = GetODataContainer();
            Order theOrder = container.Order.Expand(t => t.Customer).Where(t => t.ID == order.ID).First();
            Assert.IsNotNull(theOrder.Customer);
            Assert.AreEqual(customer.CustomerID, theOrder.Customer.CustomerID);

            container.SetLink(theOrder, "Customer", null);
            var response2 = container.SaveChanges();
            Assert.AreEqual(1, response2.Count());
            Assert.AreEqual(1, response2.Count(r => r.StatusCode == (int)HttpStatusCode.NoContent));

            container = GetODataContainer();
            Order theOrder2 = container.Order.Expand(t => t.Customer).Where(t => t.ID == order.ID).First();
            Assert.IsNull(theOrder2.Customer);
        }
    }
}

File: VB/Tests/CRUDTests.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Net
Imports [Default]
Imports NUnit.Framework
Imports WebApplication1.Models

Namespace Tests
    <TestFixture> _
    Public Class CRUDOperationTests
        Inherits ODataTestsBase

        <Test> _
        Public Sub CreateObject()
            Dim container As Container = GetODataContainer()

            Dim customer_Renamed As New Customer() With { _
                .CustomerID = "C0001", _
                .CompanyName = "Test Company" _
            }
            container.AddToCustomers(customer_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.Created), response.First().StatusCode)

            container = GetODataContainer()
            Dim createdItem As Customer = container.Customers.Where(Function(t) t.CustomerID = customer_Renamed.CustomerID).First()
            Assert.AreEqual(customer_Renamed.CustomerID, createdItem.CustomerID)
            Assert.AreEqual(customer_Renamed.CompanyName, createdItem.CompanyName)
        End Sub

        <Test> _
        Public Sub UpdateObject()
            Dim container As Container = GetODataContainer()

            Dim customer_Renamed As Customer = container.Customers.Where(Function(t) t.CustomerID = "BSBEV").First()

            customer_Renamed.CompanyName = "Test Company Renamed"
            container.UpdateObject(customer_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)

            container = GetODataContainer()
            Dim updatedItem As Customer = container.Customers.Where(Function(t) t.CustomerID = "BSBEV").First()
            Assert.AreEqual(customer_Renamed.CustomerID, updatedItem.CustomerID)
            Assert.AreEqual(customer_Renamed.CompanyName, updatedItem.CompanyName)
        End Sub

        <Test> _
        Public Sub BatchUpdate()
            Dim container As Container = GetODataContainer()
            Dim customers = container.Customers.ToList()

            For Each customer_Renamed In customers
                customer_Renamed.CompanyName &= " -renamed"
                container.UpdateObject(customer_Renamed)
            Next customer_Renamed
            Dim response = container.SaveChanges(Microsoft.OData.Client.SaveChangesOptions.BatchWithSingleChangeset)

            Assert.AreEqual(4, response.Count())
            For Each res In response
                Assert.AreEqual(CInt(HttpStatusCode.NoContent), res.StatusCode)
            Next res

            container = GetODataContainer()
            customers = container.Customers.ToList()

            For Each customer_Renamed In customers
                Assert.IsTrue(customer_Renamed.CompanyName.EndsWith(" -renamed"))
            Next customer_Renamed
        End Sub

        <Test> _
        Public Sub DeleteObject()
            Dim container As Container = GetODataContainer()

            Dim customer_Renamed As Customer = container.Customers.Where(Function(t) t.CustomerID = "WARTH").First()

            container.DeleteObject(customer_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)
        End Sub

        <Test> _
        Public Sub CreateObjectWithBlobField()
            Dim pictureData((256 * 256 * 4) - 1) As Byte
            For i As Integer = 0 To pictureData.Length - 1
                pictureData(i) = CByte(i)
            Next i
            Dim container As Container = GetODataContainer()

            Dim product_Renamed As New Product() With { _
                .ProductName = "test product", _
                .Picture = pictureData _
            }
            container.AddToProducts(product_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.Created), response.First().StatusCode)

            container = GetODataContainer()
            Dim createdItem As Product = container.Products.Where(Function(t) t.ProductName = product_Renamed.ProductName).First()

            Assert.IsNotNull(createdItem.Picture)
            Assert.AreEqual(product_Renamed.Picture.Length, createdItem.Picture.Length)
        End Sub

        <Test> _
        Public Sub UpdateObjectWithBlobField()
            Dim pictureData((256 * 256 * 4) - 1) As Byte
            For i As Integer = 0 To pictureData.Length - 1
                pictureData(i) = CByte(i)
            Next i
            Dim container As Container = GetODataContainer()

            Dim product_Renamed = container.Products.Where(Function(t) t.ProductName = "Queso Cabrales").First()
            Dim oldPicture() As Byte = product_Renamed.Picture
            product_Renamed.Picture = pictureData
            container.UpdateObject(product_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)

            container = GetODataContainer()
            Dim updatedItem As Product = container.Products.Where(Function(t) t.ProductID = product_Renamed.ProductID).First()
            Assert.IsNull(oldPicture)
            Assert.IsNotNull(updatedItem.Picture)
            Assert.AreEqual(pictureData.Length, updatedItem.Picture.Length)
        End Sub

        <Test> _
        Public Sub AddRelatedObject()
            Dim container As Container = GetODataContainer()

            Dim order_Renamed As Order = container.Orders.OrderByDescending(Function(t) t.OrderID).First()

            Dim detail As New OrderDetail() With { _
                .Quantity = 105, _
                .UnitPrice = 201.37D _
            }
            container.AddRelatedObject(order_Renamed, "OrderDetails", detail)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.Created), response.First().StatusCode)

            container = GetODataContainer()
            Dim createdItem As OrderDetail = container.OrderDetails.OrderByDescending(Function(t) t.OrderDetailID).First()
            Assert.AreEqual(detail.Quantity, createdItem.Quantity)
            Assert.AreEqual(detail.UnitPrice, createdItem.UnitPrice)
            Assert.Greater(createdItem.OrderDetailID, 0)
        End Sub

        <Test> _
        Public Sub UpdateRelatedObject()
            Dim container As Container = GetODataContainer()

            Dim order_Renamed As Order = container.Orders.Expand(Function(t) t.OrderDetails).OrderBy(Function(t) t.OrderDate).First()
            Dim detail As OrderDetail = order_Renamed.OrderDetails.OrderBy(Function(t) t.OrderDetailID).First()

            Dim oldQuantity As Short = detail.Quantity
            detail.Quantity += 1
            container.UpdateObject(detail)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)

            container = GetODataContainer()
            Dim updatedItem As OrderDetail = container.Orders.Expand(Function(t) t.OrderDetails).Where(Function(t) t.OrderID = order_Renamed.OrderID).First().OrderDetails.Where(Function(d) d.OrderDetailID = detail.OrderDetailID).First()
            Assert.AreNotEqual(oldQuantity, updatedItem.Quantity)
            Assert.AreEqual(detail.Quantity, updatedItem.Quantity)
        End Sub

        <Test> _
        Public Sub DeleteRelatedObject()
            Dim container As Container = GetODataContainer()

            Dim order_Renamed As Order = container.Orders.Expand(Function(t) t.OrderDetails).OrderBy(Function(t) t.OrderDate).First()
            Dim detail As OrderDetail = order_Renamed.OrderDetails.OrderBy(Function(t) t.OrderDetailID).First()

            container.DeleteLink(order_Renamed, "OrderDetails", detail)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)

            container = GetODataContainer()
            Dim updatedItem As Order = container.Orders.Expand(Function(t) t.OrderDetails).Where(Function(t) t.OrderID = order_Renamed.OrderID).First()
            Assert.False(updatedItem.OrderDetails.Any(Function(d) d.OrderDetailID = detail.OrderDetailID))
        End Sub

        <Test> _
        Public Sub DeleteAggregatedCollection()
            Dim container As Container = GetODataContainer()

            Dim order_Renamed As Order = container.Orders.Expand(Function(t) t.OrderDetails).OrderBy(Function(t) t.OrderDate).First()

            container.DeleteObject(order_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)

            container = GetODataContainer()
            Dim details = container.OrderDetails.Where(Function(t) t.Order.OrderID = order_Renamed.OrderID).ToList()
            Assert.AreEqual(0, details.Count)
        End Sub

        <Test> _
        Public Sub AddLink()
            Dim container As Container = GetODataContainer()

            Dim order_Renamed As Order = container.Orders.Where(Function(t) t.Customer Is Nothing).OrderByDescending(Function(t) t.OrderID).First()

            Dim customer_Renamed As Customer = container.Customers.Where(Function(t) t.CustomerID = "BSBEV").First()

            container.AddLink(customer_Renamed, "Orders", order_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)

            container = GetODataContainer()
            Dim updatedItem As Order = container.Orders.Expand(Function(t) t.Customer).Where(Function(t) t.OrderID = order_Renamed.OrderID).First()
            Assert.IsNull(order_Renamed.Customer)
            Assert.NotNull(updatedItem.Customer)
            Assert.AreEqual(updatedItem.Customer.CustomerID, customer_Renamed.CustomerID)
        End Sub

        <Test> _
        Public Sub UpdateLink()
            Dim container As Container = GetODataContainer()

            Dim order_Renamed As Order = container.Orders.Where(Function(t) t.Customer.CustomerID = "BSBEV").OrderByDescending(Function(t) t.OrderID).First()

            Dim customer_Renamed As Customer = container.Customers.Where(Function(t) t.CustomerID = "SANTG").First()

            container.AddLink(customer_Renamed, "Orders", order_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)

            container = GetODataContainer()
            Dim updatedItem As Order = container.Orders.Expand(Function(t) t.Customer).Where(Function(t) t.OrderID = order_Renamed.OrderID).First()
            Assert.AreEqual(updatedItem.Customer.CustomerID, customer_Renamed.CustomerID)
        End Sub

        <Test> _
        Public Sub DeleteLink()
            Dim container As Container = GetODataContainer()

            Dim order_Renamed As Order = container.Orders.Expand(Function(t) t.Customer).Where(Function(t) t.Customer.CustomerID = "BSBEV").OrderByDescending(Function(t) t.OrderID).First()

            container.DeleteLink(order_Renamed.Customer, "Orders", order_Renamed)
            Dim response = container.SaveChanges()

            Assert.AreEqual(1, response.Count())
            Assert.AreEqual(CInt(HttpStatusCode.NoContent), response.First().StatusCode)

            container = GetODataContainer()
            Dim updatedItem As Order = container.Orders.Expand(Function(t) t.Customer).Where(Function(t) t.OrderID = order_Renamed.OrderID).First()
            Assert.IsNotNull(order_Renamed.Customer)
            Assert.IsNull(updatedItem.Customer)
        End Sub
    End Class
End Namespace
